awsenv/replacer.go

```diff
@@ -1,89 +1,180 @@
-func (r *Replacer) ReplaceAll() (map[string]string, error) {
-	replacements := make(map[string]string)
+// Replacer handles replacing existing environment variables with values
+// retrieved from AWS Parameter Store.
+type Replacer struct {
+	ssm    ParametersGetter
+	prefix string
+}
 
-	// Get all environment vars
-	vars, err := getAllVars()
+// Apply overwrites applicable environment variables with values retrieved
+// from Parameter Store. ReplaceAll will attempt to replace as many values
+// as possible, after which it will return the first error that occurred.
+func (r *Replacer) Apply(ctx context.Context) error {
+	vars, err := r.Replacements(ctx)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to list environment")
+		return err
 	}
 
-	// Check each once for prefix
 	for name, val := range vars {
-		if strings.HasPrefix(val, r.prefix) {
-			// Get the replacement
-			newVal, err := r.replaceOne(val)
-			if err != nil {
-				return nil, errors.Wrapf(err, "failed to replace var: %s", name)
-			}
-			replacements[name] = newVal
+		suberr := setenv(name, val)
+		if err == nil && suberr != nil {
+			err = suberr
 		}
 	}
 
-	return replacements, nil
+	return err
 }
 
-func (r *Replacer) replaceOne(oldVal string) (string, error) {
-	// Trim off the prefix
-	oldVal = strings.TrimPrefix(oldVal, r.prefix)
+// Replacements returns a map of environment variable names to new values
+// that have been fetched from Parameter Store.
+func (r *Replacer) Replacements(ctx context.Context) (map[string]string, error) {
+	// param path -> env name
+	pathvars := pathmap(r.prefix, environ())
 
-	// Look up the new value in parameter store
-	out, err := r.ssm.GetParameter(&ssm.GetParameterInput{
-		Name:           aws.String(oldVal),
-		WithDecryption: aws.Bool(true),
-	})
+	// param path -> env value
+	pathvals, err := r.fetch(ctx, keys(pathvars))
 	if err != nil {
-		return "", errors.Wrapf(err, "failed to get parameter: %s", oldVal)
+		return nil, err
+	}
+
+	// env name -> env value
+	dest := make(map[string]string, len(pathvals))
+	translate(dest, pathvars, pathvals)
+
+	if len(dest) != len(pathvars) {
+		err = errors.New("missing parameters")
+	}
+
+	return dest, err
+}
+
+// ReplaceAll behaves like Replacements, but does not accept a context. To
+// directly overwrite values in the process environment, use Apply.
+func (r *Replacer) ReplaceAll() (map[string]string, error) {
+	return r.Replacements(context.Background())
+}
+
+func (r *Replacer) fetch(ctx context.Context, paths []string) (map[string]string, error) {
+
+	// eg is not ctx-wrapped, since we want to fetch as many params as
+	// possible, even if some fail. External cancellations on the passed ctx
+	// will be respected.
+	var eg errgroup.Group
+
+	decryption := true
+	batches := chunk(batchLimit, paths)
+	results := make([][]*ssm.Parameter, len(batches))
+
+	for i := range batches {
+
+		// copied to avoid race condition
+		i := i
+		batch := batches[i]
+
+		eg.Go(func() error {
```

Comment:
Are there any limits to how many concurrent requests SSM allows?

OTOH, with batches of 10, it would take a _lot_ of env vars to create that many requests.

Reply:
yeah. iirc, the default configs for aws-sdk-go do retries and may also do smart handling of throttling responses.